platforms/core-runtime/service-registry-impl/src/main/java/org/gradle/internal/service/InjectUtil.java

```diff
@@ -31,57 +36,64 @@ final class InjectUtil {
      * @param type the type to find the injectable constructor of.
      */
     static Constructor<?> selectConstructor(final Class<?> type) {
-        if (isInnerClass(type)) {
-            // The DI system doesn't support injecting non-static inner classes.
-            throw new ServiceValidationException(
-                String.format(
-                    "Unable to select constructor for non-static inner class %s.",
-                    format(type)
-                )
-            );
-        }
-
-        final Constructor<?>[] constructors = type.getDeclaredConstructors();
-        if (constructors.length == 1) {
-            Constructor<?> constructor = constructors[0];
-            if (isPublicOrPackageScoped(constructor)) {
-                // If there is a single constructor, and that constructor is public or package private we select it.
-                return constructor;
+        Constructor<?> result = CACHED_CONSTRUCTORS.get(type);
+        if (result == null) {
```

Comment:
Return early to avoid changing so much in the git-blame?

```java
Constructor<?> result = CACHED_CONSTRUCTORS.get(type);
if (result != null) {
    return result;
}

// original...
```